lib/makara/context.rb

```diff
@@ -1,83 +1,159 @@
 require 'digest/md5'
 
-# Keeps track of the current and previous context (hexdigests)
-# If a new context is needed it can be generated via Makara::Context.generate
-
+# Keeps track of the current stickiness state for different Makara configurations
 module Makara
   class Context
-    class << self
+    attr_accessor :stored_data, :staged_data
+
+    def initialize(context_data)
+      @stored_data = context_data
+      @staged_data = {}
+      @dirty = @was_dirty = false
+    end
+
+    def stage(proxy_id, ttl)
+      staged_data[proxy_id] = ttl.to_f
+    end
+
+    def stuck?(proxy_id)
+      stored_data[proxy_id] && !expired?(stored_data[proxy_id])
+    end
+
+    def staged?(proxy_id)
+      staged_data.key?(proxy_id)
+    end
 
-      def generate(seed = nil)
-        seed ||= "#{Time.now.to_i}#{Thread.current.object_id}#{rand(99999)}"
-        Digest::MD5.hexdigest(seed)
+    def release(proxy_id)
+      @dirty ||= !!stored_data.delete(proxy_id)
+      staged_data.delete(proxy_id)
+    end
+
+    def release_all
+      if self.stored_data.any?
+        self.stored_data = {}
+        # We need to track a change made to the current stored data
+        # so we can commit it later
+        @dirty = true
       end
+      self.staged_data = {}
+    end
+
+    # Stores the staged data with an expiration time based on the current time,
+    # and clears any expired entries. Returns true if any changes were made to
+    # the current store
+    def commit
+      release_expired
+      store_staged_data
+      clean
+
+      was_dirty?
+    end
+
+    private
+
+    # Indicates whether there have been changes to the context that need
+    # to be persisted when the request finishes
+    def dirty?
+      @dirty
+    end
 
-      def get_previous
-        fetch(:makara_context_previous) { generate }
+    def was_dirty?
+      @was_dirty
+    end
+
+    def expired?(timestamp)
+      timestamp <= Time.now.to_f
+    end
+
+    def release_expired
+      previous_size = stored_data.size
+      stored_data.delete_if { |_, timestamp| expired?(timestamp) }
+      @dirty ||= previous_size != stored_data.size
+    end
+
+    def store_staged_data
+      staged_data.each do |proxy_id, ttl|
+        if ttl > 0
+          self.stored_data[proxy_id] = Time.now.to_f + ttl
```

Comment:
Oh, yes! That's a really edge case (since in general `ttl` will be always the configured `master_ttl`), but it can totally happen. I'll add also a test for this.